### PR TITLE
Create IA sync endpoint

### LINKED ISSUE #7539: Create endpoint for Archive.org to update OCAIDs in OL records
### Describe the problem that you'd like solved

Today, there is no way for Internet Archive to notify Open Library of important changes to records.  If data becomes out-of-sync between the two platforms, Open Library records must be updated and synced by a human.  Before this is done, somebody has to first notice that the data is out-of-sync.  This is not ideal...

### Proposal & Constraints

Edited proposal:
Create a new POST handler and endpoint for syncing OCAID changes.  Sync requests must contain S3 credentials, edition `olid`, the original `ocaid`, and either `is_dark` if the item has been made dark, or `new_ocaid` if the OCAID has been changed.

### Stakeholders

@mekarpeles

### PR DESCRIPTION
**Pull request by jimchamp**
Closes #7539 

Creates `JSON` endpoint for IA to update OCAIDs in Open Library.  This endpoint will allow authorized IA accounts to either delete or update OCAIDs linked to editions.

### Technical

Requests to this endpoint are authenticated and validated.

#### Authentication
S3 keys are used for authentication, and are expected to be passed in the request header as `X-S3-Access` and `X-S3-Secret`.  Authorized accounts must also be members of the `/usergroup/ia` usergroup (hence the "Needs: Special Deploy" label)..

#### Validation
At a minimum, each JSON request to this endpoint must contain the following fields:
| Field | Description | Required? |
|---|---|---|
| `olid` | An edition's Open Library identifier, in the form `OL(\d+)M` | Yes |
| `action` | The operation that will be performed on the edition.  Valid values are `remove` and `modify` | Yes |
| `ocaid` | The new OCAID for an edition. | Only if `action` is `modify` |

### Special Deployment Instructions
Not a special deployment per se, but we need to add authorized IA accounts to the `/usergroup/ia` usergroup before this endpoint can be used.

### CODE REVIEW

File: openlibrary/plugins/admin/code.py [hunk 1]
```diff
@@ -239,6 +239,85 @@ def GET(self):
...
+
+        if not user or (user and not user.is_usergroup_member('/usergroup/ia')):
+            return False
+
+        return True
+
+    def validate_input(self, i):
+        """Returns True if the request is valid.
+        All requests must have an olid and an action.  If the action is
+        'modify', the requst must also include 'ocaid'.
```
Comment (jimchamp) [inline openlibrary/plugins/admin/code.py:301]:
```suggestion
        'modify', the request must also include 'ocaid'.
```

File: openlibrary/plugins/admin/code.py [hunk 2]
```diff
@@ -239,6 +239,85 @@ def GET(self):
...
+        return True
+
+    def validate_input(self, i):
+        """Returns True if the request is valid.
+        All requests must have an olid and an action.  If the action is
+        'modify', the request must also include 'ocaid'.
+        """
+        action = i.get('action', '')
+        return 'olid' in i and (
+            action == 'darken' or (action == 'modify' and 'ocaid' in i)
```
Comment (hornc) [inline openlibrary/plugins/admin/code.py:305]: `darken` is not an Open Library term (or very clear). Can we call it `remove`?

File: openlibrary/plugins/admin/code.py [hunk 3]
```diff
@@ -239,6 +239,85 @@ def GET(self):
...
+        )
+
+    def darken_record(self, edition):
+        """Deletes OCAID from given edition"""
+        data = edition.dict()
+        del data['ocaid']
+        web.ctx.site.save(data, 'Darken record')
```
Comment (hornc) [inline openlibrary/plugins/admin/code.py:312]: I'm not sure this makes sense in an OL context; the OL record is still valid. 

How about `"Remove OCAID: Item no longer available."` or similar?

and call this method `remove_ocaid` to parallel `modify_ocaid` below.

**Review by hornc — CHANGES_REQUESTED** [review]
I've left comments about the naming, which seems to be mixing domains in a confusing manner.

On further reflection on what this accomplishes, this is just removing or modifying the `ocaid` value -- don't we already have an API endpoint to do that?

Perhaps not, I can't see anything in the APIs that responds to an HTTP `PATCH`  :(

File: openlibrary/plugins/admin/code.py [hunk 4]
```diff
@@ -239,6 +239,79 @@ def GET(self):
...
+
+        if not user or (user and not user.is_usergroup_member('/usergroup/ia')):
+            return False
+
+        return True
+
+    def validate_input(self, i):
+        """Returns True if the request is valid.
+        All requests must have an olid and an action.  If the action is
+        'modify', the requst must also include 'ocaid'.
```
Comment (jimchamp) [inline openlibrary/plugins/admin/code.py:301]:
```suggestion
        'modify', the request must also include 'ocaid'.
```

File: openlibrary/plugins/admin/code.py (hunk 3 above)
Comment (jimchamp) [reply openlibrary/plugins/admin/code.py]: Thanks @hornc!  I've removed the concept of darking a record from this branch.

File: openlibrary/plugins/admin/code.py [hunk 5]
```diff
@@ -239,6 +239,85 @@ def GET(self):
...
+            case 'remove':
+                self.remove_ocaid(edition)
+            case 'modify':
+                self.modify_ocaid(edition, i.get('ocaid'))
+            case '_':
+                raise web.badrequest('Unknown action')
+
+        return delegate.RawText(json.dumps({"status": "ok"}))
+
+    @cache.memoize(
```
[the remark this answers is not in the document]
Comment (jimchamp) [reply openlibrary/plugins/admin/code.py:278]: We could do that, but it's not going to make this any safer.
How about we remove caching for now?  After this has gotten some traffic, we can determine if performance is such that caching would be helpful.

File: openlibrary/plugins/admin/code.py [hunk 6]
```diff
@@ -239,6 +239,85 @@ def GET(self):
...
+        engine="memcache",
+        key=lambda self, access, secret: f'ia_sync-{access}',
+        expires=5 * 60,
+    )
```
Comment (jimchamp) [inline openlibrary/plugins/admin/code.py:282]:
```suggestion
```
With this type of caching in place, a malicious person could modify `ocaid`s with only the S3 access key (if a genuine request was live in cache).

**Review by hornc — COMMENTED** [review]
@jimchamp , Thanks for resolving the terminology issues. This is better.

**Review by hornc — APPROVED** [review]

**Comment by mekarpeles** [comment]
LGTM as well, thank you

**Pull request merged**